File: app/src/main/java/jp/hazuki/yuzubrowser/action/item/LeftRightTabSingleAction.java
package jp.hazuki.yuzubrowser.action.item;

import android.app.AlertDialog;
import android.content.DialogInterface;
import android.os.Parcel;
import android.os.Parcelable;
import android.widget.CheckBox;

import com.fasterxml.jackson.core.JsonGenerator;
import com.fasterxml.jackson.core.JsonParser;
import com.fasterxml.jackson.core.JsonToken;

import java.io.IOException;

import jp.hazuki.yuzubrowser.R;
import jp.hazuki.yuzubrowser.action.SingleAction;
import jp.hazuki.yuzubrowser.action.view.ActionActivity;
import jp.hazuki.yuzubrowser.utils.Logger;
import jp.hazuki.yuzubrowser.utils.app.StartActivityInfo;

public class LeftRightTabSingleAction extends SingleAction {
    private static final String TAG = "LeftRightTabSingleAction";
    private static final String FIELD_NAME_TABLOOP = "0";
    private boolean mTabLoop = false;

    public LeftRightTabSingleAction(int id, JsonParser parser) throws IOException {
        super(id);

        if (parser != null) {
            if (parser.nextToken() != JsonToken.START_OBJECT) return;
            while (parser.nextToken() != JsonToken.END_OBJECT) {
                if (parser.getCurrentToken() != JsonToken.FIELD_NAME) return;
                if (!FIELD_NAME_TABLOOP.equals(parser.getCurrentName())) {
                    parser.skipChildren();
                    continue;
                }
                switch (parser.nextToken()) {
                    case VALUE_TRUE:
                        mTabLoop = true;
                        break;
                    case VALUE_FALSE:
                        mTabLoop = false;
                        break;
                    default:
                        Logger.w(TAG, "current token is not boolean value : " + parser.getCurrentToken().toString());
                        break;
                }
            }
        }
    }

    @Override
    public void writeIdAndData(JsonGenerator generator) throws IOException {
        generator.writeNumber(id);
        generator.writeStartObject();
        generator.writeBooleanField(FIELD_NAME_TABLOOP, mTabLoop);
        generator.writeEndObject();
    }

    @Override
    public int describeContents() {
        return 0;
    }

    @Override
    public void writeToParcel(Parcel dest, int flags) {
        dest.writeInt(id);
        dest.writeInt(mTabLoop ? 1 : 0);
    }

    protected LeftRightTabSingleAction(Parcel source) {
        super(source.readInt());
        mTabLoop = source.readInt() == 1;
    }

    public static final Parcelable.Creator<LeftRightTabSingleAction> CREATOR = new Parcelable.Creator<LeftRightTabSingleAction>() {
        @Override
        public LeftRightTabSingleAction createFromParcel(Parcel source) {
            return new LeftRightTabSingleAction(source);
        }

        @Override
        public LeftRightTabSingleAction[] newArray(int size) {
            return new LeftRightTabSingleAction[size];
        }
    };

    @Override
    public StartActivityInfo showSubPreference(ActionActivity context) {
        final CheckBox view = new CheckBox(context);
        view.setText(R.string.action_tab_loop);
        view.setChecked(mTabLoop);

        new AlertDialog.Builder(context)
                .setTitle(R.string.action_settings)
                .setView(view)
                .setPositiveButton(android.R.string.ok, new DialogInterface.OnClickListener() {
                    @Override
                    public void onClick(DialogInterface dialog, int which) {
                        mTabLoop = view.isChecked();
                    }
                })
                .setNegativeButton(android.R.string.cancel, null)
                .show();

        return null;
    }

    public boolean isTabLoop() {
        return mTabLoop;
    }

}

File: legacy/src/main/java/jp/hazuki/yuzubrowser/legacy/action/item/LeftRightTabSingleAction.kt
/*
 * Copyright (C) 2017-2019 Hazuki
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

package jp.hazuki.yuzubrowser.legacy.action.item

import android.app.AlertDialog
import android.os.Parcel
import android.os.Parcelable
import android.view.View
import android.widget.CheckBox
import com.squareup.moshi.JsonReader
import com.squareup.moshi.JsonWriter
import jp.hazuki.yuzubrowser.legacy.R
import jp.hazuki.yuzubrowser.legacy.action.SingleAction
import jp.hazuki.yuzubrowser.legacy.action.view.ActionActivity
import jp.hazuki.yuzubrowser.ui.app.StartActivityInfo
import java.io.IOException

class LeftRightTabSingleAction : SingleAction {
    var isTabLoop = false
        private set

    @Throws(IOException::class)
    constructor(id: Int, reader: JsonReader?) : super(id) {
        if (reader != null) {
            if (reader.peek() != JsonReader.Token.BEGIN_OBJECT) return
            reader.beginObject()
            while (reader.hasNext()) {
                if (reader.peek() != JsonReader.Token.NAME) return
                when (reader.nextName()) {
                    FIELD_NAME_TAB_LOOP -> {
                        if (reader.peek() != JsonReader.Token.BOOLEAN) return
                        isTabLoop = reader.nextBoolean()
                    }
                    else -> reader.skipValue()
                }
            }
            reader.endObject()
        }
    }

    @Throws(IOException::class)
    override fun writeIdAndData(writer: JsonWriter) {
        writer.value(id)
        writer.beginObject()
        writer.name(FIELD_NAME_TAB_LOOP)
        writer.value(isTabLoop)
        writer.endObject()
    }

    override fun describeContents(): Int {
        return 0
    }

    override fun writeToParcel(dest: Parcel, flags: Int) {
        dest.writeInt(id)
        dest.writeInt(if (isTabLoop) 1 else 0)
    }

    private constructor(source: Parcel) : super(source.readInt()) {
        isTabLoop = source.readInt() == 1
    }

    override fun showSubPreference(context: ActionActivity): StartActivityInfo? {
        val view = View.inflate(context, R.layout.action_checkbox, null)
        val checkBox = view.findViewById(R.id.checkBox) as CheckBox
        checkBox.setText(R.string.action_tab_loop)
        checkBox.isChecked = isTabLoop

        AlertDialog.Builder(context)
                .setTitle(R.string.action_settings)
                .setView(view)
            .setPositiveButton(android.R.string.ok) { _, _ -> isTabLoop = checkBox.isChecked }
                .setNegativeButton(android.R.string.cancel, null)
                .show()

        return null
    }

    companion object {
        private const val TAG = "LeftRightTabSingleAction"
        private const val FIELD_NAME_TAB_LOOP = "0"

        @JvmField
        val CREATOR: Parcelable.Creator<LeftRightTabSingleAction> = object : Parcelable.Creator<LeftRightTabSingleAction> {
            override fun createFromParcel(source: Parcel): LeftRightTabSingleAction {
                return LeftRightTabSingleAction(source)
            }

            override fun newArray(size: Int): Array<LeftRightTabSingleAction?> {
                return arrayOfNulls(size)
            }
        }
    }

}
